cpmpy/solvers/pumpkin.py

```diff
@@ -0,0 +1,550 @@
+
+    ===============
+    Installation
+    ===============
+
+    The `pumpkin_solver_py` python package is currently not available on PyPI.
+    It can be installed from source using the following steps:
+     1. clone the repository from github: https://github.com/consol-lab/pumpkin
+     2. install the "maturin" package to build the python bindings: $ pip install maturin
+     3. build and install the package: :code:`cd pumpkin/pumpkin-solver-py && maturin develop`
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_pumpkin
+
+    ==============
+    Module details
+    ==============
+"""
+import warnings
+import re
+
+from os.path import join
+
+import numpy as np
+
+from cpmpy.exceptions import NotSupportedError
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..expressions.core import Expression, Comparison, Operator, BoolVal
+from ..expressions.globalconstraints import GlobalConstraint
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl, intvar, boolvar
+from ..expressions.utils import is_num, is_any_list, get_bounds
+from ..transformations.get_variables import get_variables
+from ..transformations.linearize import canonical_comparison
+from ..transformations.normalize import toplevel_list
+from ..transformations.decompose_global import decompose_in_tree
+from ..transformations.flatten_model import flatten_constraint
+from ..transformations.comparison import only_numexpr_equality
+from ..transformations.reification import reify_rewrite, only_bv_reifies, only_implies
+from ..transformations.safening import no_partial_functions
+
+import time
+
+
+class CPM_pumpkin(SolverInterface):
+    """
+    Interface to Pumpkin's API
+
+    Creates the following attributes (see parent constructor for more):
+    - ``pum_solver``: the pumpkin.Model() object
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import pumpkin_solver_py as gp
+            return True
+        except ImportError:
+            return False
+
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: None, not used
+        """
+        if not self.supported():
+            raise Exception("CPM_Pumpkin: Install the python package 'pumpkin_solver_py'")
+
+        from pumpkin_solver_py import Model
+
+        assert subsolver is None 
+
+        # initialise the native solver object
+        self.pum_solver = Model() 
+
+        # for objective
+        self._objective = None
+        self.objective_is_min = True
+
+        # initialise everything else and post the constraints/objective
+        super().__init__(name="pumpkin", cpm_model=cpm_model)
+
+    @property
+    def native_model(self):
```

Comment:
do we do that for other solvers? there its just in the docs that it is called self.pum_solver, and we don't provide an indirect/abstraction?

Reply:
We have it for other solvers too, not sure when it was introduced, but already quite some time I think